File: 0x00 基本算法/0x03 前缀和与差分/AcWing99.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 5010;
int n, m;
int s[N][N];

int main() {
  int cnt, R;
  cin >> cnt >> R;
  R = min(5001, R); // 前缀和从1开始，横纵坐标 +1 后最大格子坐标 5001

  while (cnt -- ) {
    int x, y, w;
    cin >> x >> y >> w;
    x ++ , y ++ ; // 前缀和习惯从 1 开始，不用处理边界
    s[x][y] += w;
  }

  // 预处理二维前缀和：s[i][j] 表示 (1,1) 到 (i,j) 的矩形价值总和
  n = m = 5001;
  for (int i = 1; i <= n; i ++ )
    for (int j = 1; j <= m; j ++ )
      s[i][j] += s[i - 1][j] + s[i][j - 1] - s[i - 1][j - 1];

  // 枚举边长为 R 的正方形的右下角坐标 (i, j)
  int res = 0;
  for (int i = R; i <= n; i ++ )
    for (int j = R; j <= m; j ++ )
      res = max(res, s[i][j] - s[i - R][j] - s[i][j - R] + s[i - R][j - R]);

  cout << res << endl;
  return 0;
}

File: 0x00 基本算法/0x03 前缀和与差分/AcWing99.py
import sys

def main():
    input_data = sys.stdin.read().split()
    if not input_data:
        return
        
    iterator = iter(int(x) for x in input_data)
    cnt = next(iterator)
    R = min(5001, next(iterator))
    
    size = 5002
    s = [[0] * size for _ in range(size)]
    
    for _ in range(cnt):
        x = next(iterator) + 1
        y = next(iterator) + 1
        w = next(iterator)
        s[x][y] += w
        
    for i in range(1, 5002):
        row = s[i]
        prev_row = s[i - 1]
        for j in range(1, 5002):
            row[j] += prev_row[j] + row[j - 1] - prev_row[j - 1]
            
    res = 0
    for i in range(R, 5002):
        row = s[i]
        prev_row = s[i - R]
        for j in range(R, 5002):
            val = row[j] - prev_row[j] - row[j - R] + prev_row[j - R]
            if val > res:
                res = val
                
    print(res)

if __name__ == '__main__':
    main()